### PR TITLE
[14.0][LOC] l10n_ch: Add SCOR (iso) reference type for swiss qr invoices

### PR DESCRIPTION
**Pull request by TeoGoddet**
Description of the issue/feature this PR addresses:
According to https://www.paymentstandards.ch/en/shared/know-how/faq/qr.html
The following types can be used:

QR-bill with QR IBAN and QR reference
QR-bill with IBAN and creditor reference (ISO 11649)
QR-bill with IBAN without reference
The difference lies in the use of IBAN and references.

Current behavior before PR:
QR-bill with IBAN and creditor reference (ISO 11649) is not supported yet and downgraded as "without reference".

Desired behavior after PR is merged:
QR-bill with IBAN and creditor reference (ISO 11649) are included in the qr code if they are present.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_ch/models/res_bank.py [hunk 1]
```diff
@@ -191,6 +192,10 @@ def _l10n_ch_get_qr_vals(self, amount, currency, debtor_partner, free_communicat
             # _check_for_qr_code_errors ensures we can't have a QR-IBAN without a QR-reference here
             reference_type = 'QRR'
             reference = structured_communication
+        elif self._is_creditor_reference(structured_communication):
```
Comment (lordkrandel) [inline addons/l10n_ch/models/res_bank.py:195]: We could remove the `@staticmethod` s and have 

```suggestion
        # The structured communication can also be a valid creditor reference (ISO 11649)
        # reference: https://github.com/arthurdejong/python-stdnum/blob/master/stdnum/iso11649.py
        elif iso11649.is_valid(structured_communication):
            reference_type = 'SCOR'
            reference = iso11649.validate(structured_communication)
```

**Review by lordkrandel — CHANGES_REQUESTED** [review]
There is some preliminary check on the structured_communication that also does not consider the Creditor Reference case.
Is the IBAN + Creditor Reference mandatory or is the `no reference` case an allowed one in our case?

https://github.com/odoo/odoo/blob/68166794f005f1e47dbd2cf8b9fe7ffe0be3ff66/addons/l10n_ch/models/res_bank.py#L318

**Comment by TeoGoddet** [comment]
@lordkrandel I'm not sure to understand, 
When using normal IBAN, you can either have
  - no reference (only free communication)
  - Creditor Reference

when using QR iban you need : 
 - QRR/ISR reference (27 chars)

https://github.com/odoo/odoo/blob/68166794f005f1e47dbd2cf8b9fe7ffe0be3ff66/addons/l10n_ch/models/res_bank.py#L318
So, i think this check is great for us and not a problem.

**Comment by TeoGoddet** [comment]
resolves #71578

File: addons/l10n_ch/models/res_bank.py [hunk 2]
```diff
@@ -191,6 +192,11 @@ def _l10n_ch_get_qr_vals(self, amount, currency, debtor_partner, free_communicat
             # _check_for_qr_code_errors ensures we can't have a QR-IBAN without a QR-reference here
             reference_type = 'QRR'
             reference = structured_communication
+        # The structured communication can also be a valid creditor reference (ISO 11649)
+        # reference: https://github.com/arthurdejong/python-stdnum/blob/master/stdnum/iso11649.py
+        elif iso11649.is_valid(structured_communication):
+            reference_type = 'SCOR'
+            reference = iso11649.validate(structured_communication)
```
Comment (oco-odoo) [inline addons/l10n_ch/models/res_bank.py:199]: Shouldn't you rather call compact(), so that you don't run the verification twice? (is_valid() calls validate())

File: addons/l10n_ch/models/res_bank.py [hunk 3]
```diff
@@ -191,6 +192,11 @@ def _l10n_ch_get_qr_vals(self, amount, currency, debtor_partner, free_communicat
             # _check_for_qr_code_errors ensures we can't have a QR-IBAN without a QR-reference here
             reference_type = 'QRR'
             reference = structured_communication
+        # The structured communication can also be a valid creditor reference (ISO 11649)
+        # reference: https://github.com/arthurdejong/python-stdnum/blob/master/stdnum/iso11649.py
```
Comment (oco-odoo) [inline addons/l10n_ch/models/res_bank.py:196]: Remove this link. We can see that you're calling stdnum from the code, and the line above is already making clear what you're checking exactly.

File: doc/cla/individual/TeoGoddet.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+Switzerland, 2021-12-11
```
Comment (oco-odoo) [inline doc/cla/individual/TeoGoddet.md:1]: This should be in another commit, with the [CLA] tag https://www.odoo.com/documentation/15.0/developer/misc/other/guidelines.html#tag-and-module-name

**Review by oco-odoo — CHANGES_REQUESTED** [review]
Also, your commit message does not follow the guidelines ( https://www.odoo.com/documentation/15.0/developer/misc/other/guidelines.html#commit-message-structure ), please adapt it.

**Comment by TeoGoddet** [comment]
Hi ! I applied all requests from review
But i have an error of stdnum, maybe installed stdnum is too old, i'm checking

image

File: addons/l10n_ch/models/res_bank.py [hunk 5]
```diff
@@ -3,6 +3,12 @@
 
 import re
 
+try:
+    from stdnum import iso11649
+except ImportError:
+    import stdnum
+    _logger.warning("stdnum.iso11649 is avalaible from stdnum >= 1.10. The one installed is %s. You cannot use l10n_ch creditor reference." % stdnum.__version__)
```
Comment (lordkrandel) [inline addons/l10n_ch/models/res_bank.py:10]: `avalaible` -> `available`

Comment (lordkrandel) [reply addons/l10n_ch/models/res_bank.py:10]: Also, _logger is undefined there. 

See:
https://github.com/odoo/odoo/blob/a7f7233e0eae8ee101d745a9813cba930fd03dcb/addons/l10n_ar/models/res_partner_bank.py#L4-L12

Comment (lordkrandel) [reply addons/l10n_ch/models/res_bank.py:10]: That code in `l10n_ar` also creates an alternative `validate_cbu` function to be used and handles the case
https://github.com/odoo/odoo/tree/master/addons/l10n_ar/models/res_partner_bank.py#L43-L49
while here we just log a warning. 🤔

Comment (TeoGoddet) [reply addons/l10n_ch/models/res_bank.py:13]: @lordkrandel I don't know what is the best solution, but i can ship the needed functions (is_valid and compact) if you prefer

Comment (oco-odoo) [reply addons/l10n_ch/models/res_bank.py]: The version of stdnum that is in the requirements in inferior to 1.10 (1.8), so indeed, since it's only 2 functions we need, it'd probably be better to add them directly in the module, and not rely on stdnum for now.

**Comment by oco-odoo** [comment]
@TeoGoddet your commit is missing the [IMP] label in front of its title ;)

File: addons/l10n_ch/models/res_bank.py (hunk 5 above)
Comment (TeoGoddet) [reply addons/l10n_ch/models/res_bank.py]: removed dependency on sdtnum

**Comment by TeoGoddet** [comment]
I've also corrected the report to display it correctlyas Reference and not unstructured communication

File: addons/l10n_ch/report/swissqr_report.xml [hunk 6]
```diff
@@ -57,15 +57,23 @@
...
+                                    <span class="title">Reference</span>
                                 </div>
+                            </t>
+                            <t t-if="o.partner_bank_id._is_qr_iban()">
                                 <div class="swissqr_text content">
                                     <span t-esc="o.space_qrr_reference(o.payment_reference)"/><br/>
                                     <br/>
                                 </div>
                             </t>
+                            <t t-if="o.partner_bank_id._is_iso11649_reference()">
```
Comment (oco-odoo) [inline addons/l10n_ch/report/swissqr_report.xml:71]: Instead of calling _is_iso11649_reference() and _is_qr_iban, I suggest using 2 t-set to store the results of these calls and reuse them as variables.

File: addons/l10n_ch/models/res_bank.py [hunk 7]
```diff
@@ -286,6 +290,16 @@ def _is_qr_reference(self, reference):
...
 
+    @api.model
+    def _is_iso11649_reference(self, reference):
+        """ Checks whether the given reference is a ISO11649 (SCOR) reference.
+        """
+        return reference \
+               and len(reference) >= 5 \
+               and len(reference) <= 25 \
+               and reference.startswith('RF') \
+               and stdnum.iso7064.mod_97_10.validate(reference[4:] + reference[:4])
```
Comment (oco-odoo) [inline addons/l10n_ch/models/res_bank.py:301]: My version of stdnum (the one from the requirements) doesn't contain iso7064 and crashes here.

File: addons/l10n_ch/report/swissqr_report.xml [hunk 8]
```diff
@@ -57,15 +57,23 @@
...
+                                <div class="swissqr_text content">
+                                    <span t-esc="o.space_iso11649_reference(o.payment_reference)"/><br/>
```
Comment (oco-odoo) [inline addons/l10n_ch/report/swissqr_report.xml:73]: That function doesn't exist; I guess you renamed it.

File: addons/l10n_ch/report/swissqr_report.xml [hunk 9]
```diff
@@ -57,15 +57,23 @@
                                 <br/>
                             </div>
 
-                            <t t-if="o.partner_bank_id._is_qr_iban()">
+                            <t t-if="o.partner_bank_id._is_qr_iban() or o.partner_bank_id._is_iso11649_reference()">
                                 <div class="swissqr_text title">
-                                    <span>Reference</span>
+                                    <span class="title">Reference</span>
```
Comment (oco-odoo) [inline addons/l10n_ch/report/swissqr_report.xml:62]: Why adding the class here ? Doesn't seem to change anything.

**Review by oco-odoo — CHANGES_REQUESTED** [review]
Crashes for multiple reasons. You didn't test :p

File: addons/l10n_ch/report/swissqr_report.xml (hunk 8 above)
Comment (TeoGoddet) [reply addons/l10n_ch/report/swissqr_report.xml]: sorry I missed up

File: addons/l10n_ch/models/res_bank.py (hunk 7 above)
Comment (TeoGoddet) [reply addons/l10n_ch/models/res_bank.py]: removed dependency on iso7064

File: addons/l10n_ch/report/swissqr_report.xml (hunk 6 above)
Comment (TeoGoddet) [reply addons/l10n_ch/report/swissqr_report.xml]: changed it

File: addons/l10n_ch/report/swissqr_report.xml [hunk 10]
```diff
@@ -57,15 +57,23 @@
                                 <br/>
                             </div>
 
-                            <t t-if="o.partner_bank_id._is_qr_iban()">
+                            <t t-if="o.partner_bank_id._is_qr_iban() or o.partner_bank_id._is_iso11649_reference()">
```
[the remark this answers is not in the document]
Comment (TeoGoddet) [reply addons/l10n_ch/report/swissqr_report.xml]: sorry

**Comment by TeoGoddet** [comment]
@oco-odoo sorry for this, i was not understanding why the red. Thanks !

**Review by oco-odoo — APPROVED** [review]
@robodoo r+

**Comment by oco-odoo** [comment]
@robodoo rebase-ff
